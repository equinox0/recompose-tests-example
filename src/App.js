import PropTypes from "prop-types";
import React, { Fragment } from "react";
import { compose, lifecycle, withHandlers, withState } from "recompose";
import { getVehicle } from "./api/vehicle";
import { Loader, SearchForm, Vehicle } from "./components";

const App = ({ isLoading, onSearchValuesChange, onSubmitSearch, searchValues, vehicle }) => (
    <Fragment>
        <h1>Star wars vehicle</h1>
        <SearchForm
            onChange={onSearchValuesChange}
            onSubmit={onSubmitSearch}
            values={searchValues}
        />
        {isLoading ? (
            <Loader />
        ) : vehicle ? (
            <Vehicle vehicle={vehicle} />
        ) : (
            <div>Brak pojazdu o podanym Id</div>
        )}
    </Fragment>
);

App.propTypes = {
    isLoading: PropTypes.bool.isRequired,
    onSearchValuesChange: PropTypes.func.isRequired,
    onSubmitSearch: PropTypes.func.isRequired,
    searchValues: PropTypes.shape({
        vehicleId: PropTypes.number.isRequired
    }).isRequired,
    vehicle: PropTypes.shape({
        model: PropTypes.string,
        name: PropTypes.string
    })
};

const enhance = compose(
    withState("isLoading", "setIsLoading", false),
    withState("searchValues", "setSearchValues", { vehicleId: 0 }),
    withState("vehicle", "setVehicle", undefined),
    withHandlers({
        onSearchValuesChange: ({ setSearchValues }) => values => {
            setSearchValues(values);
        },
        onSubmitSearch: ({ searchValues, setIsLoading, setVehicle }) => async () => {
            setIsLoading(true);
            try {
                const vehicle = await getVehicle(searchValues.vehicleId);
                setVehicle(vehicle);
            } catch (error) {
                setVehicle();
            } finally {
                setIsLoading(false);
            }
        }
    }),
    lifecycle({
        async componentDidMount() {
            this.props.onSubmitSearch();
        }
    })
);

export { enhance, App };
export default enhance(App);
