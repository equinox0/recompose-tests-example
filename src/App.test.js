import Enzyme from "enzyme";
import React from "react";
import { App, enhance } from "./App";

describe("<App/>", () => {
    describe("base component", () => {
        it("should show test text", () => {
            const wrapper = Enzyme.shallow(<App showText />);
            expect(wrapper.contains(<div>Test text</div>)).toEqual(true);
        });

        it("show not show test text", () => {
            const wrapper = Enzyme.shallow(<App />);
            expect(wrapper.contains(<div>Test text</div>)).not.toEqual(true);
        });
    });

    describe("enhancer", () => {
        it("should have props - testOne", () => {
            const MockComponent = () => null;
            const EnhancedComponent = enhance(MockComponent);
            const wrapper = Enzyme.mount(<EnhancedComponent />);
            const props = wrapper.find(MockComponent).props();
            expect(props.showText).toEqual(true);
        });
    });
});
