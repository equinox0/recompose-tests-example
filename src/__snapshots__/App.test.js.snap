// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> component should show loader 1`] = `
<Fragment>
  <h1>
    Star wars vehicle
  </h1>
  <withHandlers(SearchForm) />
  <div>
    Brak pojazdu o podanym Id
  </div>
</Fragment>
`;

exports[`<App/> component should show vehicle if it exists 1`] = `
<Fragment>
  <h1>
    Star wars vehicle
  </h1>
  <withHandlers(SearchForm) />
  <Vehicle
    vehicle={
      Object {
        "model": "Vehicle model",
        "name": "Vehicle name",
      }
    }
  />
</Fragment>
`;
